Eurotronic Spirit: When `heatsetpoint` is changed, set `mode` to `auto`.

tkintscher opened the pull request:
See #1098.

manup commented on rest_sensors.cpp:
```diff
@@ -1364,6 +1364,24 @@ int DeRestPluginPrivate::changeSensorConfig(const ApiRequest &req, ApiResponse &
+                            hostFlags = val.value.u32;
+                        }
+
+			hostFlags &= 0xffffeb;
```
https://github.com/dresden-elektronik/deconz-rest-plugin/issues/1098#issuecomment-462077343

Please add a small comment  to annote the related flags, after a few months it's hard to tell what they do :)

// clear set off (0x10) and boost (0x04) flags
hostFlags &= ~(0x10 | 0x04);

tkintscher replied:
you're right, it is more clear this way. Done :)

ebaauw commented:
This won’t work, _Off_ isn’t a single bit.  you need to set 0x000010 to clear 0x000010 (and set 0x000020 to set 0x000010).

tkintscher commented:
Right, thanks. I've changed it and checked that setting `heatsetpoint` sets the _Host Flags_ back to 0x01.

ebaauw commented:
I didn't dare to use ~0x04, since the Zigbee value is u24 and the hostFlags is quint32.  That's why I hardcoded the inverse.  Also I specified the full 24-bit value.

Looking at the full file, I see `hostFlags` already gets written - good.  However, this might cause `config.displayflipped` and `config.locked` to be cleared.  You need to retrieve the current value of `hostFlags` from `RConfigHostFlags` (as when handling `RConfigMode` etc).  The ZCL value is 0x000000 after starting deCONZ, until the first _Report Attributes_ (or _Read Attributes Response_) from the thermostat has been processed.  `RConfigHosts` is persisted in the database and restored on startup.

tkintscher commented:
> I didn't dare to use ~0x04, since the Zigbee value is u24 and the hostFlags is quint32. That's why I hardcoded the inverse. Also I specified the full 24-bit value.

Yes, I saw that and was torn between both options. I've thought ~0x04 would be more readable than 0xfb.

> Looking at the full file, I see `hostFlags` already gets written - good. However, this might cause `config.displayflipped` and `config.locked` to be cleared. You need to retrieve the current value of `hostFlags` from `RConfigHostFlags` (as when handling `RConfigMode` etc). The ZCL value is 0x000000 after starting deCONZ, until the first _Report Attributes_ (or _Read Attributes Response_) from the thermostat has been processed. `RConfigHosts` is persisted in the database and restored on startup.

I do check whether hostFlags has been set before and retrieve the value if necessary, see the full change:
https://github.com/dresden-elektronik/deconz-rest-plugin/pull/1246/files
Just tested it and `locked` and `displayflipped` are kept.

ebaauw commented:
> Just tested it and `locked` and `displayflipped` are kept.

Try this shortly after restarting deCONZ.

tkintscher commented:
I've changed it, but I cannot test it because restarting deCONZ sets the mode to `auto` anyway (locally and on the device).

I also removed the event on `config.mode` as per your suggestion.

ebaauw commented:
Thanks.

> restarting deCONZ sets the mode to `auto` anyway (locally and on the device).

It doesn't for me.  Double-checked the REST API values and the thermostat's display.  Note that the deCONZ GUI will reflect the default values (2000 for _Heat Set Point_ 0x4003 and 0x000000 for _Host Flags_ 0x4008) until the first _Report Attributes_ or _Read Attributes Response_ has been received.

ebaauw commented:
@manup, could you merge this PR as well?  Thanks.

manup commented:
Sure :) thanks guys

ebaauw commented:
Final test: it works as expected:
```
$ ph put /sensors/8/config '{"mode": "heat"}'
# restart deCONZ
$ ph get /sensors/8
{
  "config": {
    "battery": 95,
    "displayflipped": true,
    "heatsetpoint": 3000,
    "locked": false,
    "mode": "heat",
    "offset": 0,
    "on": true,
    "reachable": true
  },
  "ep": 1,
  "etag": "39d2c5c4a9db37ff2fce83f98561d7d8",
  "manufacturername": "Eurotronic",
  "modelid": "SPZB0001",
  "name": "Living Room Radiator",
  "state": {
    "lastupdated": "2019-02-12T20:37:28",
    "on": true,
    "temperature": 2238,
    "valve": 155
  },
  "swversion": "15181120",
  "type": "ZHAThermostat",
  "uniqueid": "00:15:8d:00:01:92:d2:51-01-0201"
}
$ ph put /sensors/8/config '{"heatsetpoint": 2150}'
# wait for it...
$ ph get /sensors/8
{
  "config": {
    "battery": 95,
    "displayflipped": true,
    "heatsetpoint": 2150,
    "locked": false,
    "mode": "auto",
    "offset": 0,
    "on": true,
    "reachable": true
  },
  "ep": 1,
  "etag": "d7f9e96a8505e7f871e7cab5f94ee27c",
  "manufacturername": "Eurotronic",
  "modelid": "SPZB0001",
  "name": "Living Room Radiator",
  "state": {
    "lastupdated": "2019-02-12T20:38:17",
    "on": true,
    "temperature": 2238,
    "valve": 255
  },
  "swversion": "15181120",
  "type": "ZHAThermostat",
  "uniqueid": "00:15:8d:00:01:92:d2:51-01-0201"
}
```
Web socket notifications:
```
Feb 12 21:37:14 pi1 dc_eventlog[792]: /sensors/8/config: {"mode":"heat","reachable":true}
Feb 12 21:37:21 pi1 dc_eventlog[792]: /sensors/8/config: {"heatsetpoint":3000}
Feb 12 21:37:21 pi1 dc_eventlog[792]: /sensors/8/state: {"lastupdated":"2019-02-12T20:37:21"}
Feb 12 21:37:21 pi1 dc_eventlog[792]: /sensors/8/state: {"lastupdated":"2019-02-12T20:37:21","temperature":2238}
Feb 12 21:37:28 pi1 dc_eventlog[792]: /sensors/8/state: {"lastupdated":"2019-02-12T20:37:28","valve":155}
Feb 12 21:37:32 pi1 dc_eventlog[792]: connection closed
Feb 12 21:37:42 pi1 dc_eventlog[792]: listening on ws://localhost:443
Feb 12 21:37:52 pi1 dc_eventlog[792]: /sensors/8/state: {"lastupdated":"2019-02-12T20:37:52","valve":255}
Feb 12 21:38:11 pi1 dc_eventlog[792]: /sensors/8/config: {"heatsetpoint":2150,"reachable":true}
Feb 12 21:38:17 pi1 dc_eventlog[792]: /sensors/8/config: {"mode":"auto"}
Feb 12 21:38:17 pi1 dc_eventlog[792]: /sensors/8/state: {"lastupdated":"2019-02-12T20:38:17"}
Feb 12 21:38:26 pi1 dc_eventlog[792]: /sensors/8/state: {"lastupdated":"2019-02-12T20:38:26","valve":129}
```
